src/systems/advanced_lift_drag/AdvancedLiftDrag.cc

```diff
@@ -0,0 +1,825 @@
+
+  /*
+  Next, get the properties of each control surface: which joint it connects to,
+  which direction it deflects in, and the effect of its deflection on the
+  coefficient of drag, side force, lift, roll moment, pitching moment, and yaw moment.
+  */
+
+  while( _sdf->HasElement("control_surface") )
+  {
+    auto curr_ctrl_surface = _sdf->GetElement("control_surface");
+    auto ctrl_surface_name = curr_ctrl_surface->GetElement(
+      "name")->Get<std::string>();
+    auto entities =
+        entitiesFromScopedName(ctrl_surface_name, _ecm, this->model.Entity());
+
+    if (entities.empty())
+    {
+      gzerr << "Joint with name[" << ctrl_surface_name << "] not found. "
+             << "The LiftDrag will not generate forces\n";
+      this->validConfig = false;
+      return;
+    }
+    else if (entities.size() > 1)
+    {
+      gzwarn << "Multiple joint entities with name[" << ctrl_surface_name
+              << "] found. Using the first one.\n";
+    }
+
+    controlJointEntity = *entities.begin();
+    if (!_ecm.EntityHasComponentType(this->controlJointEntity,
+                                     components::Joint::typeId))
+    {
+      this->controlJointEntity = kNullEntity;
+      gzerr << "Entity with name[" << ctrl_surface_name << "] is not a joint\n";
+      this->validConfig = false;
+      return;
+    }
+
+    this->controlJoints.push_back(controlJointEntity);
+    this->ctrl_surface_direction.push_back(
+      std::stod(((curr_ctrl_surface->GetElement(
+        "direction"))->GetValue())->GetAsString()));
+    this->CD_ctrl.push_back(
+      std::stod(((curr_ctrl_surface->GetElement(
+        "CD_ctrl"))->GetValue())->GetAsString()));
+    this->CY_ctrl.push_back(
+      std::stod(((curr_ctrl_surface->GetElement(
+        "CY_ctrl"))->GetValue())->GetAsString()));
+    this->CL_ctrl.push_back(
+      std::stod(((curr_ctrl_surface->GetElement(
+        "CL_ctrl"))->GetValue())->GetAsString()));
+    this->Cell_ctrl.push_back(
+      std::stod(((curr_ctrl_surface->GetElement(
+        "Cell_ctrl"))->GetValue())->GetAsString()));
+    this->Cem_ctrl.push_back(
+      std::stod(((curr_ctrl_surface->GetElement(
+        "Cem_ctrl"))->GetValue())->GetAsString()));
+    this->Cen_ctrl.push_back(
+      std::stod(((curr_ctrl_surface->GetElement(
+        "Cen_ctrl"))->GetValue())->GetAsString()));
+    _sdf->RemoveChild(curr_ctrl_surface);
+  }
+
+  this->AR = _sdf->Get<double>("AR", this->AR).first;
+  this->mac = _sdf->Get<double>("mac", this->mac).first;
+  this->eff = _sdf->Get<double>("eff", this->eff).first;
+  this->rho = _sdf->Get<double>("air_density", this->rho).first;
+  this->radialSymmetry = _sdf->Get<bool>("radial_symmetry",
+      this->radialSymmetry).first;
+  this->area = _sdf->Get<double>("area", this->area).first;
+
+  // blade forward (-drag) direction in link frame
+  this->forward =
+      _sdf->Get<math::Vector3d>("forward", this->forward).first;
+  this->forward.Normalize();
```

Comment:
Knit: Check length is not zero and complain if it is. `Normalize()` does not complain if lengths are zero.